source/reduce/structured_loop_to_selection_reduction_opportunity.cpp

```diff
@@ -0,0 +1,406 @@
+  }
+
+  if (merge_inst->opcode() == SpvOpLoopMerge) {
+    // In the case of a loop we have the further requirement that, to be in
+    // the construct, the block must not be dominated by the loop's continue
+    // target.
+    if (dominator_analysis->Dominates(
+            merge_inst->GetSingleWordOperand(kContinueNodeIndex), block_id)) {
+      return false;
+    }
+  }
+
+  // The SPIR-V spec's definition of structured control flow is unnatural for
+  // various scenarios where loop breaks and continues are nested inside
+  // conditionals.  To account for this in the short term, we impose the
+  // additional restriction that the block must appear, in program order,
+  // before the construct's merge block in order to be part of the construct.
+  //
+  // TODO: This is a workaround that should be replaced in due course with a
+  // more precise check if the rules for structured control flow are refined.
+  for (auto& bb : *enclosing_function_) {
+    if (bb.id() == block_id) {
+      // We find block_id before the merge block, so we decide that block_id
+      // is part of the structured control flow construct.
+      return true;
+    }
+    if (bb.id() == merge_block_id) {
+      // We find the merge block before block_id.  This suggests that block_id
+      // should not really be part of the construct.
+      return false;
+    }
+  }
+  return true;
+}
+
+void StructuredLoopToSelectionReductionOpportunity::RedirectEdge(
+    uint32_t source_id, uint32_t original_target_id, uint32_t new_target_id) {
+  assert(source_id != original_target_id);
+  assert(source_id != new_target_id);
+  assert(original_target_id != new_target_id);
+
+  // Redirect the edge; depends on what kind of branch instruction is involved.
+  auto terminator = context_->cfg()->block(source_id)->terminator();
+  if (terminator->opcode() == SpvOpBranch) {
+    assert(terminator->GetSingleWordOperand(0) == original_target_id);
+    terminator->SetOperand(0, {new_target_id});
+  } else {
+    assert(terminator->opcode() == SpvOpBranchConditional);
+    assert(terminator->GetSingleWordOperand(1) == original_target_id ||
+           terminator->GetSingleWordOperand(2) == original_target_id);
+    for (auto index : {1u, 2u}) {
+      if (terminator->GetSingleWordOperand(index) == original_target_id) {
+        terminator->SetOperand(index, {new_target_id});
+      }
+    }
+  }
+
+  // The old and new targets may have phi instructions; these will need to
+  // respect the change in edges.
+  AdaptPhiNodesForRemovedEdge(source_id,
+                              context_->cfg()->block(original_target_id));
+  AdaptPhiNodesForAddedEdge(source_id, context_->cfg()->block(new_target_id));
+}
+
+void StructuredLoopToSelectionReductionOpportunity::AdaptPhiNodesForRemovedEdge(
+    uint32_t from_id, BasicBlock* to_block) {
+  for (auto& inst : *to_block) {
+    if (inst.opcode() != SpvOpPhi) {
+      // Phi instructions must appear first in a block, so if we find a non-phi
```

Comment:
These can be interleaved with OpLine instructions. This assumes debug instructions have been stripped.

Reply:
I've added a test for this case, but it turns out that the code is OK as is since OpLine information is attached to other instructions rather than being preserved in the IR.